Upgrade maps without parsing ruleset

pchote opened the pull request:
Creating a `Map` object throws an exception if the mod's default Ruleset is not valid.  This is a good thing in most cases, but is a major blocker for #14600 - even an innocuous error (from a yaml syntax perspective) like a missing required key is enough to completely break the upgrade process.

I expect that @reaperrr, @abcdefg30, and others will be familiar with this frustration, but for completeness I have included a testcase that demonstrates the issue when cherry-picked to bleed.

This sidesteps the crash by applying the upgrade rules directly to the map yaml without first parsing a `Map`.  This has an extra side-effect of significantly improving the upgrade rule run time.

pchote committed:
Apply map upgrade rules without instantiating the Map object.

This avoids a crash that stops the maps from being updated
when the base mod ruleset is not valid (e.g. a trait is missing a
FieldLoader.Required property).  This also significantly improves the
upgrade performance.

abcdefg30 commented on OpenRA.Mods.Common/UtilityCommands/UpgradeMapCommand.cs:
```diff
@@ -65,13 +75,29 @@ public static void UpgradeMap(ModData modData, IReadWritePackage package, int en
-			UpgradeRules.UpgradePlayers(modData, engineDate, ref map.PlayerDefinitions, null, 0);
-			UpgradeRules.UpgradeActors(modData, engineDate, ref map.ActorDefinitions, null, 0);
-			map.Save(package);
+			var yaml = new MiniYaml(null, MiniYaml.FromStream(package.GetStream("map.yaml"), package.Name));
```
This will crash with an NRE later on if there is no `map.yaml` file present in the directory. Can we get a proper error message here instead? (Or ignore such maps/folders.)

abcdefg30 reviewed (dismissed):
Looks good to me otherwise.

pchote commented:
Good catch, fixed.

MunWolf commented on OpenRA.Mods.Common/UtilityCommands/UpgradeMapCommand.cs:
```diff
@@ -38,16 +48,16 @@ static void ProcessYaml(ModData modData, Map map, MiniYaml yaml, int engineDate,
 					// HACK: Obtain the writable save path using knowledge of the underlying filesystem workings
 					var packagePath = filename;
-					var package = map.Package;
+					var p = package;
```
Any reason why you don't use the package parameter directly here instead of the p variable?

pchote replied:
This is inside a loop, and each file needs access to the original `package` value.

pchote commented:
Rebased and removed test case.  I count 2 :+1:'s counting my own implicit one, but I would prefer @abcdefg30 or another @OpenRA/engine-hackers to re-approve instead of overriding the merge approval.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/3bf03e6acae3d6f9720011cab1e8705bb00e351b...0de24b8c811802f7415de3d4c6c16be995a3686d)
